src/main/java/com/myFirstProject/myFirstProject/service/BasketServiceImpl.java

```diff
@@ -39,18 +39,28 @@
     public Long saveBasket(BasketReq basketReq) {
         logger.info("Get order to update {}", basketReq);
         sumValidation(basketReq);
+        promoCodeValidation(basketReq);
         Basket order = reqConverterService.convert(basketReq);
         LocalDateTime time = LocalDateTime.now();
         order.setExpiredTime(time.plusDays(1L));
         Basket order1 = basketRepository.save(order);
         Long id = order1.getId();
         logger.info("Basket saved with id {}", id);
-        accountService.payForArticles(basketReq.getSum(), basketReq.getUser().getId());
+        accountService.payForArticles(basketReq.getSum(), basketReq.getUser().getId(), basketReq.getPromoCode());
+        basketReq.getPromoCode().setValid(false);
         logger.info(String.format("%s was payed for order", basketReq.getSum()));
 
         return id;
     }
 
+    private void promoCodeValidation(BasketReq basketReq) {
+        if (!basketReq.getPromoCode().getValid()){
+            throw new PromoCodeNotValidException(String.format("PromoCode with id - %s have been used", basketReq.getId()));
+        }else if (basketReq.getPromoCode().getExpiredDate().compareTo(LocalDateTime.now())<0){
```

Comment:
code format. again it is very strange that you trust client instead of checking values from DB